File: mona_ticker.pl
#!/home/naoto/.plenv/shims/perl

use strict;
use warnings;
use URI;
use LWP::UserAgent;
use utf8;
use JSON qw/encode_json decode_json/;
use Encode;
use Email::MIME;
use Email::MIME::Creator;
use Email::Sender::Simple 'sendmail';
use Email::Sender::Transport::SMTP;



#ask,bidのしきい値と、現在のbtc価格を設定
#しきい値を超えるとメール通知が行われます
my $ask=16;
my $bid=10;
my $btc_jpy=65300;

while(1){
        call_ticker();
	sleep(1800); #指定秒数後に再実行
}


sub call_ticker{

my $date;
my %tmp;

#$tmp　に各サイトの値（ask,bid)を保存
$tmp{"monatr"}=	&monatr();#もなとれ価格は別に表示


#api取得先アドレスを配列で複数指定
my %api_add=(
	"etwings" => "https://exchange.etwings.com/api/1/ticker/mona_jpy",
	"monax" => "https://monax.jp/api/pricemncjpyv2",
	);


#LWPで接続し、価格情報JSONをGETする
foreach my $key (keys %api_add){
	my $uri=URI->new($api_add{$key});
	my $ua = LWP::UserAgent->new();
	my $res= $ua-> get($uri);
	warn $res->status_line if $res->is_error;

#受け取ったJSONデータをパース
my $json=$res->content;
my $ref=decode_json($json); #Perlのデータ構造にデコード変換


print "$key => ask:$ref->{ask}  bid:$ref->{bid}\n";

#%tmpに保存
$tmp{$key}={ask=>$ref->{ask},bid=>$ref->{bid}};

}

#最後に%tmpの各ask,bidがしきい値を満たすか調べ、満たす場合はsend_mail

foreach my $key (keys %tmp){
    if(($tmp{$key}{ask}>=$ask) || ($tmp{$key}{bid}<=$bid)){

		&send_mail(\%tmp); #メール送信
		last;
    }
}
	print "\n\n";
}


sub monatr{

	my $uri=URI->new("https://api.monatr.jp/ticker");

#取得したい通貨ペアを指定

	$uri->query_form(
		market=> "MONA_BTC"
		);

	my $ua = LWP::UserAgent->new();
	my $res= $ua-> get($uri);
	warn $res->status_line if $res->is_error;

	#JSONをパース
	my $json=$res->content;
	my $ref=decode_json($json);
	my $cur_ask=$ref->{current_ask}*$btc_jpy;
	my $cur_bid=$ref->{current_bid}*$btc_jpy;
	my $key="monatr";

	#$tmp{$key}={ask=>$cur_ask,bid=>$cur_bid};

	printf ("monatr=> ask:%.1f  bid:%.1f \n",$cur_ask,$cur_bid);
	
	return {ask=>$cur_ask,bid=>$cur_bid};

}



sub send_mail{
    my $ref_tmp=shift;
    my $from=q/"monatick" <送信元アドレス>/;
    my $to=q/"送信先名" <送信先アドレス>/;
    my $subject=q/価格情報のお知らせ　From mona_ticker/;
    my $text;

    $text="monatickからmonacoin価格情報のお知らせです。\n\n\n";

	foreach my $key(keys %$ref_tmp){
              #print "$key \n";
	    $text.="$key=>  ask:$ref_tmp->{$key}->{ask}  bid:$ref_tmp->{$key}->{bid} \n";
        }

my $email = Email::MIME->create(
  header => [
    From    => encode('MIME-Header-ISO_2022_JP' => $from),
    To      => encode('MIME-Header-ISO_2022_JP' => $to),
    Subject => encode('MIME-Header-ISO_2022_JP' => $subject ),
  ],
  attributes => {
    content_type => 'text/plain',
    charset      => 'ISO-2022-JP',
    encoding     => '7bit',
  },
  body => encode('iso-2022-jp' =>

		 "$text"

),

);

#自分の設定したSMTPサーバからメールを送信
my $smtp = Email::Sender::Transport::SMTP->new({
               'host'          => 'SMTPサーバのアドレス',
               'port'          => ポート番号,
               'sasl_username' => 'ユーザ名＊＊＊＊＊',
               'sasl_password' => 'パスワード＊＊＊＊＊＊',
           });

sendmail($email, { 'transport' => $smtp });

}

File: mona_ticker.php
<?php
require_once "Mail.php";

//ask,bidのしきい値と、現在のbtc価格を設定
$target_ask=18; //この値以上で通知
$target_bid=10; //この値以下で通知

$btc_jpy=55300;//BtcPrice in Japan


while(1){
	call_ticker($target_ask,$target_bid,$btc_jpy);
	sleep(60); #10秒ごとに実行する
}


function call_ticker($target_ask,$target_bid,$btc_jpy){

//&monatr($btc_jpy,$ask,$bid);#もなとれ価格は別に表示

//apiアドレスを配列で複数指定
$api_add = array(
	"etwings" => "https://exchange.etwings.com/api/1/ticker/mona_jpy",
	//"もなっくす" => "https://monax.jp/api/pricemncjpyv2",
	);


//価格情報を取得する
foreach ($api_add as $key =>$value){
	$res=http_get($value);
	//print($res);
	preg_match('/{.*\}/', $res,$json);
	$json=json_decode($json[0]);
	$ask=$json->ask;
	$bid=$json->bid;

	print "$key  ask:$ask bid:$bid \n";

	if($ask>$target_ask || $bid<$target_bid){
		send_mail($key,$ask,$bid);
	}



}
}

function send_mail($key,$ask,$bid){
//日本語メールを送る際に必要
mb_language("Japanese");
mb_internal_encoding("UTF-8");


$sendto="＊＊＊＊";  //あなたのメール送信先
$sendfrom="＊＊＊＊"; //メール送信元（自由）

// SMTPサーバーの情報を連想配列にセット
$params = array(
  "host" => "＊＊＊＊＊＊",   // SMTPサーバー名
  "port" => ＊＊＊,              // ポート番号
  "auth" => true,            // SMTP認証を使用する
  "username" => "＊＊＊＊＊",  // SMTPのユーザー名
  "password" => "＊＊＊＊＊＊"   // SMTPのパスワード
);

// PEAR::Mailのオブジェクトを作成
// ※バックエンドとしてSMTPを指定
$mailObject = Mail::factory("smtp", $params);


// 送信先のメールアドレス
$recipients = "$sendto";

// メールヘッダ情報を連想配列としてセット
$headers = array(
  "To" => "$sendto",         // →ここで指定したアドレスには送信されない
  "From" => "$sendfrom",
  "Subject" =>  mb_encode_mimeheader("mona_tickerからの通知") // 日本語の件名を指定する場合、mb_encode_mimeheaderでエンコード
);

// メール本文
$body="mona_tickerからの通知\n\n $key  買値:$ask 売値:$bid \n";

// sendメソッドでメールを送信
$mailObject->send($recipients, $headers, $body);
}

?>